packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -74,324 +186,476 @@ class FocusNode extends ChangeNotifier {
+    final FocusScopeNode previousEnclosingScope = child.enclosingScope;
+    final bool hadChildren = _children.isNotEmpty;
+    super.reparentIfNeeded(childContext, child);
+    final FocusScopeNode currentEnclosingScope = child.enclosingScope;
+    // If the child moved scopes, then the policy data is invalid.
+    if (previousEnclosingScope != currentEnclosingScope) {
+      policyData = null;
+    }
+    // If we just added our first child to this scope, and this scope had the
+    // focus, then focus the child, otherwise make it the current focus.
+    if (!hadChildren && currentEnclosingScope.focusedChild == null && currentEnclosingScope.hasFocus) {
+      child.requestFocus();
+    }
+  }
+
+  /// Returns true if this scope is the focused child of its parent scope.
+  bool get isFirstFocus => enclosingScope.focusedChild == this;
+
+  /// Add the given node as a child of this focus scope, and make it be the
+  /// [focusedChild].
+  void setFirstFocus(BuildContext childContext, FocusScopeNode node) {
     assert(node != null);
-    if (_focus != node)
-      return;
-    _focus._parent = null;
-    _focus._manager = null;
-    _focus = null;
-    _didChangeFocusChain();
+    if (node._parent == null) {
+      reparentIfNeeded(childContext, node);
+    }
+    assert(node.ancestors.contains(this));
+    node.requestFocus();
   }
 
-  /// Makes the given child the first focus of this scope.
+  /// If this scope lacks a focus, reparent the node as a child of this one, and
+  /// request that the given node becomes the focus.
   ///
-  /// If the child has another parent scope, the child is first removed from
-  /// that scope. After this method returns [isFirstFocus] will be true for
-  /// the child.
-  void setFirstFocus(FocusScopeNode child) {
-    assert(child != null);
-    if (_firstChild == child)
-      return;
-    child.detach();
-    _prepend(child);
-    assert(child._parent == this);
-    _didChangeFocusChain();
+  /// Useful for widgets that wish to grab the focus if no other widget already
+  /// has the focus.
+  ///
+  /// The node is notified that it has received the overall focus in a
+  /// microtask.
+  void autofocus(BuildContext context, FocusNode node) {
+    if (focusedChild == null) {
+      if (node._parent == null) {
+        reparentIfNeeded(context, node);
+      }
+      assert(node.ancestors.contains(this));
+      node.requestFocus();
+    }
   }
 
-  /// Adopts the given scope if it is the first focus of another scope.
+  /// Returns the child of this node which should receive focus if this scope
+  /// node receives focus.
   ///
-  /// A widget that sets a scope as the first focus of another scope should
-  /// call this method during its `build` method in case the widget is moved
-  /// from one location in the tree to another location that has a different
-  /// focus scope.
+  /// If [hasFocus] is true, then this points to the child of this node which is
```

Comment:
this points to the child of this node which is currently focused  => the returned node has the focus.